Update spack-stack docs for better user onboarding

eap opened the pull request:
Update documents to have more straightforward entry points for users of pre-configured clusters and for local developers looking to create a new environment.


## Issues Address

- Fixes: #488

eap committed:
Reorganize code

eap committed:
add using spack-stack section

eap committed:
Improving wording, no tests yet

eap committed:
overview.rst change

eap committed:
merge

eap committed:
fix index name of NewSiteConfigs

eap committed:
several rst bugs and small pre-review edits

eap committed:
remove quickstart.rst other small fixes

eap commented on doc/source/Platforms.rst:
```diff
@@ -477,517 +477,174 @@ For ``spack-stack-1.2.0``/``skylab-3.0.0``, use a c6i.2xlarge instance or simila
    module available
 
-..  _Platform_New_Site_Configs:
+.. _Platforms_NewEnv:
```
The git diff on this is terrible and you aught to view this as a monolith change not a line-by-line change. What happened here is that the "New Site Configs" section was moved to its own document. The mildly similar "Installing a new spack environment" section was moved out of the quick-start and put here.

eap replied:
You can improve the diff substantially by reviewing changes from the first commit to head, rather than from the branching point. In my first commit I reorganized the files and did bulk text relocation, so the diff from there shows most of my edits (although some reorganizing was done in later commits, so it isn't perfect).

eap commented on doc/source/Platforms.rst:
```diff
@@ -477,517 +477,174 @@ For ``spack-stack-1.2.0``/``skylab-3.0.0``, use a c6i.2xlarge instance or simila
-Unlike in previous versions, the instructions below assume that ``Python`` is built by ``spack``. That means that when using the ``spack`` environments (i.e., loading the modules for building or running code), the ``spack`` installation of ``Python`` with its available ``Python`` modules should be used to ensure consistency. However, a Homebrew ``Python`` installation may still be needed to build ...
-
-It is recommended to not use ``mpich`` or ``openmpi`` installed by Homebrew, because these packages are built using a flat namespace that is incompatible with the JEDI software. The spack-stack installations of ``mpich`` and ``openmpi`` use two-level namespaces as required.
+=====================================
+Installing a new spack environment
+=====================================
```
It is unclear to me this section belongs here (the tail end of platforms), but it seemed more applicable here than in New Site Configs which caters more to end-to-end work as opposed to preconfigured site work which is referenced extensively in this section.

eap commented on doc/source/Platforms.rst:
```diff
@@ -477,517 +477,174 @@ For ``spack-stack-1.2.0``/``skylab-3.0.0``, use a c6i.2xlarge instance or simila
 
-   definitions:
-   - compilers: ['%gcc']
+=================================================
```
Should this be the first section or is it ok here as the last section?

eap reviewed (commented):
initial self review with some comments.

eap commented on doc/source/BuildingContainers.rst:
```diff
@@ -0,0 +1,40 @@
+
+Container recipes are self-contained except that at container creation time (``spack stack create ctr ...``) package information from the common ``packages.yaml`` (or a manually provided version) is added to the ``packages:`` section, and specs are added in the section ``specs`` as defined in the template (default is an empty template with no specs.) Container recipes are defined in ``configs/con ...
+
+**WORK IN PROGRESS**
```
This section was moved from its original location in Environments.rst. Some of the meta comments were removed and consolidated with the paragraph above, but this is still a work in progress so I left the warning.

climbfuji reviewed (changes requested):
A bit of an unusual PR review ... I read through the final documentation and put my thoughts together in one place. There are two sections, the first is about this PR  itself, the second about future work/note to self.

**This PR**

- rename "Platforms" to "Pre-configured sites" or "Pre-configured platforms"
- rename "Pre-configured sites" to "Fully supported spack-stack installations"?
- Where is the table in 2.1.1?
- rename "Installing a new spack environment" to "Installing a new spack environment on a pre-configured site"
- remove subsection 2.3.1 - all this can be under 2.3 directly
- Revisit (I had to remove this several times to fix annoying problems when checking out code) in 3.1.3. If it's needed, it also needs to go into the Ubuntu 20, 22, and Red Hat 8 sections for consistency. Or be removed from there if it's not needed.
```
# git-lfs must configure system git hooks.
git lfs install
```
- NEVER do! This changes with every spack-stack installation. The only time this needs to be set is when building environments with spack-stack/spack. For using environments, this is not needed. Also counter the earlier warning/recommendation in the documentation to put all sorts of stuff in your `.bash_profile` (this file should remain unchanged except for very few exceptions). This needs to be removed from both the macOS and Linux section.
```
Add this SPACK_STACK_DIR to your shell environment.
echo "export SPACK_STACK_DIR=$(pwd)" >> $HOME/.bash_profile
```
- Accordingly, revert this (both the macOS and Linux section):
```
module use $SPACK_STACK_DIR/
```
- Sections "3.1.5. Using your environment to build code" and "3.2.5. Using your environment to build code" are basically or less duplicates, and the instructions in 3.2.5 to use `lmod` are not correct, since the current Linux instructions have `tcl` environment modules for Red Hat 8 and Ubuntu 20, but `lmod` for Ubuntu 22. I feel that these sections could be generalized and combined (3.1.5, 3.2.5, 2.3) into one separate section (top-level section before "Platforms/Pre-configured sites" and "Generating new site configs"; this is what MOST users will be interested in - not how to build, but how to use).
- Move section [6.1. Manual software installations](https://spack-stack--499.org.readthedocs.build/en/499/Prerequisites.html#manual-software-installations) into section 9, Maintainer/developer section as new section 9.1 (everything else shifts). Section 6 will then be empty and can be removed.

**Future work/note to self**

This needs to be updated (outside of the scope of this PR) before the next release:
- Section overview, list of maintainers/developers
- Maybe turn 2.1.1 into an intro paragraph + table for 2.1 and then start 2.1.1 orion etc ...
- Update all pre-configured sites in 2.1.x
- Split into pre-configured/fully-supported/tier-1 and tier-2 (site config exists, but we don't support fully and rely on users/community to provide updates)
- Think about moving "2.3. Using a spack environment to compile and run code" elsewhere (but see above for changes in this PR)
- "Section 4.1.1. Container build example" - add info on converting Docker containers into singularity at the end. Consider removing "Work in progress".
- Consider removing section 5 entirely until someone has time to redo it (don't delete, just remove from the list of included files for generating the documentation)
- Update 9.4. Recommended Directory Layout - this is no longer correct, we discussed a better directory layout in the spack-stack meetings

climbfuji commented:
@eap I updated the target branch for this PR to the newly created release/1.3.0 branch and updated your PR with the latest changes from that branch.

climbfuji committed:
Merge branch 'release/1.3.0' into bugfix/eap-docs

srherbener commented on doc/source/NewSiteConfigs.rst:
```diff
@@ -0,0 +1,585 @@
+Prerequisites (one-off)
+-----------------------
+
+This instructions are meant to be a reference that users can follow to set up their own system. Depending on the user's setup and needs, some steps will differ, some may not be needed and others may be missing. Also, the package versions may change over time.
```
This instructions ... ==> These instructions ...

eap replied:
done

srherbener commented on doc/source/NewSiteConfigs.rst:
```diff
@@ -0,0 +1,585 @@
+   source setup.sh
+
+   # Add this SPACK_STACK_DIR to your shell environment.
+   echo "export SPACK_STACK_DIR=$(pwd)" >> $HOME/.bash_profile
```
I agree with @climbfuji that this is not a good thing to do. For example, if you decide to move your spack-stack clone/installation you have to remember to update your .bash_profile and restart your bash session.

eap replied:
done

srherbener reviewed (commented):
Overall looks great! I like the reorganization which makes it easier to find everything you need when starting with your laptop. I just had a couple minor comments.

eap committed:
Rename platforms and fix all links

eap committed:
fix table?

eap committed:
many changes

eap committed:
fix index

eap committed:
review comments

eap committed:
Move prerequisites to doc/source/MaintainersSection.rst

eap commented:
Using quote reply to track some of this work.

> * rename "Platforms" to "Pre-configured sites" or "Pre-configured platforms"
Done

> * rename "Pre-configured sites" to "Fully supported spack-stack installations"?
Clarification needed?

> * Where is the table in 2.1.1?
Interesting, this is missing in the current docs as well. Somehow the table got misformatted. I've fixed it in the current PR.

> * rename "Installing a new spack environment" to "Installing a new spack environment on a pre-configured site"
Done

> * remove subsection 2.3.1 - all this can be under 2.3 directly
I think this was "pre-configured sites" under "using a spack environment to compile and run code". This has been reorganized into a new section #2 "Using spack-stack environments" which serves as a short quick start and consolidates the redundant "using spack stack" clauses at the end of Mac setup, Linux setup, and preconfigured environments.

> * `git lfs install` in 3.1.3. Revisit If it's needed, it also needs to go into the Ubuntu 20, 22, and Red Hat 8 sections for consistency. Or be removed from there if it's not needed.
Removed per our discussion.

> (bashrc, bash_profile change) * NEVER do! This changes with every spack-stack installation. The only time this needs to be set is when building environments with spack-stack/spack. For using environments, this is not needed. Also counter the earlier warning/recommendation in the documentation to put all sorts of stuff in your `.bash_profile`.
Removed.

> * Sections "3.1.5. Using your environment to build code" and "3.2.5. Using your environment to build code" are basically or less duplicates, and the instructions in 3.2.5 to use `lmod` are not correct, since the current Linux instructions have `tcl` environment modules for Red Hat 8 and Ubuntu 20, but `lmod` for Ubuntu 22. I feel that these sections could be generalized and combined (3.1.5, 3.2.5, 2.3) into one separate section (top-level section before "Platforms/Pre-configured sites" and "Generating new site configs"; this is what MOST users will be interested in - not how to build, but how to use).

New section created "Using Spack Environments" is now section 2 and combines these sections. The old sections now end with whatever specific information is needed and a link to section 2.

> * Move section [6.1. Manual software installations](https://spack-stack--499.org.readthedocs.build/en/499/Prerequisites.html#manual-software-installations) into section 9, Maintainer/developer section as new section 9.1 (everything else shifts). Section 6 will then be empty and can be removed.

Done

climbfuji reviewed (changes requested):
This is really coming along! I noticed a few things, but then we can put this child to bed:
- New section two has two missing/malformatted references/hyperlinks to other sections
- I would rename "3.1 spack-stack-v1" to "3.1 Officially supported spack-stack installations" (sorry for not being clear in my first review). Thanks for fixing the table.
- There is a rogue section 3.16.1 that can be removed (it's empty)

srherbener reviewed (approved):
Thanks for addressing my comments, and again thanks for improving this documentation!

eap committed:
review comments

eap commented:
PTAL

climbfuji committed:
Merge branch 'release/1.3.0' into bugfix/eap-docs

climbfuji reviewed (approved):
Thanks for workingon this and for addressing my comments. So much better than before!
